internal/state/relationship/capturer.go

```diff
@@ -3,50 +3,70 @@ package relationship
 import (
 	v1 "k8s.io/api/core/v1"
 	discoveryV1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/gateway-api/apis/v1beta1"
 
 	"github.com/nginxinc/nginx-kubernetes-gateway/internal/manager/index"
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/graph"
 )
 
 //go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Capturer
 
 // Capturer captures relationships between Kubernetes objects and can be queried for whether a relationship exists
 // for a given object.
 //
-// Currently, it only captures relationships between HTTPRoutes and Services and Services and EndpointSlices,
-// but it can be extended to capture additional relationships.
 // The relationships between HTTPRoutes -> Services are many to 1,
 // so these relationships are tracked using a counter.
 // A Service relationship exists if at least one HTTPRoute references it.
-// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+// An EndpointSlice relationship exists if its Service owner is referenced by at least one HTTPRoute.
+//
+// A Namespace relationship exists if it has labels that match a Gateway listener's label selector.
 type Capturer interface {
 	Capture(obj client.Object)
-	Remove(resourceType client.Object, nsname types.NamespacedName)
-	Exists(resourceType client.Object, nsname types.NamespacedName) bool
+	Remove(resource client.Object, nsname types.NamespacedName)
+	Exists(resource client.Object, nsname types.NamespacedName) bool
 }
 
 type (
 	// routeToServicesMap maps HTTPRoute names to the set of Services it references.
 	routeToServicesMap map[types.NamespacedName]map[types.NamespacedName]struct{}
 	// serviceRefCountMap maps Service names to the number of HTTPRoutes that reference it.
 	serviceRefCountMap map[types.NamespacedName]int
+	// gatewayLabelSelectorsMap maps Gateways to the label selectors that their listeners use for allowed routes
+	gatewayLabelSelectorsMap map[types.NamespacedName][]labels.Selector
+	// namespaceCfg holds information about a namespace
+	// - labels that it contains
+	// - gateways that reference it (if labels match)
+	// - whether or not it matches a listener's label selector
+	namespaceCfg struct {
+		labelMap map[string]string
+		gateways map[types.NamespacedName]struct{}
+		match    bool
+	}
+	// referencedNamespaces is a collection of namespaces in the system
+	referencedNamespaces map[types.NamespacedName]namespaceCfg
```

Comment:
are those namespaces are referencedNamespaces (as in Gateway reference them) or all namespaces in the cluster? If the case is the latter, I suggest renaming the field to reflect that.

Reply:
It's all namespaces, I can update the name.